helix-term/src/commands/dap.rs

```diff
@@ -0,0 +1,856 @@
+            for (breakpoint, dap_breakpoint) in breakpoints.iter_mut().zip(dap_breakpoints) {
+                breakpoint.id = dap_breakpoint.id;
+                breakpoint.verified = dap_breakpoint.verified;
+                breakpoint.message = dap_breakpoint.message;
+                // TODO: handle breakpoint.message
+                // TODO: verify source matches
+                breakpoint.line = dap_breakpoint.line.unwrap_or(0).saturating_sub(1); // convert to 0-indexing
+                                                                                      // TODO: no unwrap
+                breakpoint.column = dap_breakpoint.column;
+                // TODO: verify end_linef/col instruction reference, offset
+            }
+        }
+        Err(e) => anyhow::bail!("Failed to set breakpoints: {}", e),
+        _ => {}
+    };
+    Ok(())
+}
+
+pub fn dap_toggle_breakpoint_impl(cx: &mut Context, path: PathBuf, line: usize) {
+    // TODO: need to map breakpoints over edits and update them?
+    // we shouldn't really allow editing while debug is running though
+
+    let breakpoints = cx.editor.breakpoints.entry(path.clone()).or_default();
+    // TODO: always keep breakpoints sorted and use binary search to determine insertion point
+    if let Some(pos) = breakpoints
+        .iter()
+        .position(|breakpoint| breakpoint.line == line)
+    {
+        breakpoints.remove(pos);
+    } else {
+        breakpoints.push(Breakpoint {
+            line,
+            ..Default::default()
+        });
+    }
+
+    let debugger = match &mut cx.editor.debugger {
+        Some(debugger) => debugger,
+        None => return,
+    };
+
+    if let Err(e) = breakpoints_changed(debugger, path, breakpoints) {
+        cx.editor
+            .set_error(format!("Failed to set breakpoints: {}", e));
+    }
+}
+
+pub fn dap_continue(cx: &mut Context) {
+    let debugger = match &mut cx.editor.debugger {
+        Some(debugger) => debugger,
+        None => return,
+    };
+
+    if let Some(thread_id) = debugger.thread_id {
+        let request = debugger.continue_thread(thread_id);
+        if let Err(e) = block_on(request) {
+            cx.editor.set_error(format!("Failed to continue: {}", e));
+            return;
+        }
+        resume_application(debugger);
+    } else {
+        cx.editor
+            .set_error("Currently active thread is not stopped. Switch the thread.".into());
+    }
+}
+
+pub fn dap_pause(cx: &mut Context) {
+    thread_picker(cx, |editor, thread| {
+        let debugger = match &mut editor.debugger {
+            Some(debugger) => debugger,
+            None => return,
+        };
+        let request = debugger.pause(thread.id);
+        // NOTE: we don't need to set active thread id here because DAP will emit a "stopped" event
+        if let Err(e) = block_on(request) {
+            editor.set_error(format!("Failed to pause: {}", e));
+        }
+    })
+}
+
+pub fn dap_step_in(cx: &mut Context) {
```

Comment:
`step_in`, `step_out`, `next`, etc all seem to share most of the code, differing only in the `let request = debugger.step_in()` part.

Reply:
Not much we can do here since the future types are all different :/ only thing I can think of is a macro to generate these